### PR TITLE
ENH: Adding pd.__git_version__ to point to git sha commit (#21295)

### LINKED ISSUE #21295: access git revision of currently running pandas
I would like to know from inside the code which pandas git revision I am running at the moment. So I can easily confirm I am on proper branch/revision when testing some new features/bug fixes.
Is it possible in pandas? If it isn't I would like this issue to be FR for such functionality. Just git commit hash in enough.
In dask you can access this info with `dask.__git_revision__`.

### PR DESCRIPTION
**Pull request by atulagrwl**
- [ ] closes #21295 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/__init__.py [hunk 1]
```diff
@@ -83,6 +83,7 @@
 from ._version import get_versions
 v = get_versions()
 __version__ = v.get('closest-tag', v['version'])
+__git_version__ = v['full-revisionid']
```
Comment (jreback) [inline pandas/__init__.py]: this won’t work for released version
use .get

pls add a test for this

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/__init__.py (hunk 1 above)
Comment (atulagrwl) [reply pandas/__init__.py]: Adding test to check if it returns valid sha should be okay?

**Comment by jreback** [comment]
we have a once a day CI - https://github.com/pandas-dev/pandas-ci
that does a full install via pip and conda and then runs tests like a user
i believe this will fail here (these are post merge tests)

as __git_version__ would be None

can u verify?

File: pandas/tests/test_common.py [hunk 2]
```diff
@@ -262,3 +264,9 @@ def test_compression_warning(compression_only):
                                         check_stacklevel=False):
             with f:
                 df.to_csv(f, compression=compression_only)
+
+
+def test_git_version():
+    git_version = pd.__git_version__
```
Comment (jreback) [inline pandas/tests/test_common.py]: can u add the issue number here

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a release note in 0.24 other enhancements

**Comment by atulaggarwal** [comment]
> we have a once a day CI - https://github.com/pandas-dev/pandas-ci
that does a full install via pip and conda and then runs tests like a user
i believe this will fail here (these are post merge tests)
I am not sure how to do that. This is my first commit to pandas and would appreciate if you tell me help me.

**Comment by atulagrwl** [comment]
Merged latest master to resolve conflict in whatsnew file.

**Comment by atulaggarwal** [comment]
@jbrockmendel - Did the rebase and minor refactoring as suggested in test.

**Comment by atulagrwl** [comment]
@jbrockmendel - Did the rebase and minor refactoring as suggested in test.

File: doc/source/whatsnew/v0.24.0.txt [hunk 3]
```diff
@@ -180,6 +180,8 @@ Other Enhancements
 - :func:`~DataFrame.to_csv` and :func:`~DataFrame.to_json` now support ``compression='infer'`` to infer compression based on filename (:issue:`15008`)
 - :func:`to_timedelta` now supports iso-formated timedelta strings (:issue:`21877`)
 - :class:`Series` and :class:`DataFrame` now support :class:`Iterable` in constructor (:issue:`2193`)
+- New method :meth:`__git_version__` will return git commit sha of current build
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.24.0.txt]: This is an attribute, not a method, right?

**Comment by jbrockmendel** [comment]
@jreback looks about right.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/__init__.py [hunk 4]
```diff
@@ -83,6 +83,7 @@
 from ._version import get_versions
 v = get_versions()
 __version__ = v.get('closest-tag', v['version'])
+__git_version__ = v.get('full-revisionid')
```
[the remark this answers is not in the document]
Comment (atulaggarwal) [reply pandas/__init__.py]: Is there a way I can create a built version locally and test it out? @jreback

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply pandas/__init__.py]: python setup.py bdist_wheel, and then install that wheel in a new env.

[the remark this answers is not in the document]
Comment (atulaggarwal) [reply pandas/__init__.py]: Created wheel from above command (https://www.dropbox.com/s/3ieyl07ctcsja9u/pandas-0.24.0.dev0%2B373.g942172546-cp36-cp36m-macosx_10_7_x86_64.whl?dl=0) and it showed git_version correctly

```
>>> import pandas as pd
>>> pd.__version__
'0.24.0.dev0+373.g942172546'
>>> pd.__git_version__
'9421725461061b55bc84e388d3952957d9b0cd83'
```

File: doc/source/whatsnew/v0.24.0.txt [hunk 5]
```diff
@@ -180,6 +180,8 @@ Other Enhancements
 - :func:`~DataFrame.to_csv` and :func:`~DataFrame.to_json` now support ``compression='infer'`` to infer compression based on filename (:issue:`15008`)
 - :func:`to_timedelta` now supports iso-formated timedelta strings (:issue:`21877`)
 - :class:`Series` and :class:`DataFrame` now support :class:`Iterable` in constructor (:issue:`2193`)
+- New attribute :attr:`__git_version__` will return git commit sha of current build
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: add the issue ref here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by alimcmaster1** [comment]
@atulagrwl this is almost there, mind merging in the latest changes to these files? And updating the issue ref as per @jreback comment.

**Comment by alimcmaster1** [comment]
Im happy to take this over the line, since we haven't heard back from @atulagrwl . Will reopen a new PR and we can close this. 

CC. @jangorecki

File: pandas/tests/test_common.py [hunk 6]
```diff
@@ -165,3 +167,10 @@ def test_compression_warning(compression_only):
                                         check_stacklevel=False):
             with f:
                 df.to_csv(f, compression=compression_only)
+
+
+def test_git_version():
+    # GH 21295
+    git_version = pd.__git_version__
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/tests/test_common.py]: does this actually run when pandas is installed after tagging?
iow need to tag (locally) then build a release and install and pd.test()

[the remark this answers is not in the document]
Comment (alimcmaster1) [reply pandas/tests/test_common.py]: Looks like it will, I ran:
python setup.py bdist_wheel
and installed that wheel in a fresh virtual env.

**Comment by jschendel** [comment]
Closing in favor of #22745

**Pull request closed without merging**